metabase/metabase
# Add custom warning message when leaving editing a model (query definition or metadata)

Closes #33793

### Description
- This PR is only about editing models. If you open up a model, click "show editor" and change something there - this is not model editing - it is ad-hoc question creation. To edit a model you need to choose "Edit model query" or "Edit metadata" (see demo).
- History is unreliable when editing a model. That's why in scenarios below I suggest to press and hold browser back button and navigate as far as possible.
- Showing a warning when clicking "Cancel" while editing a model will be handled in a separate issue: #34366
- Similar to:
  - #34084 
  - #34192
  - #33827
  - #34272

### How to verify

Prerequisites for all scenarios:

1. Create a model (either native query or notebook) and save it
2. Open the model

**1. Custom warning modal is shown when leaving editing model query with unsaved changes**

1. Click "Edit query definition"
2. Change something in the query definition
3. Press and hold back button in your browser to open up history menu
4. Go back in history further than when you opened the model

Expected result: custom warning modal **is shown**.

**2. Custom warning modal is shown when leaving editing model metadata with unsaved changes**

1. Click "Edit metadata"
2. Change something in the metadata
3. Press and hold back button in your browser to open up history menu
4. Go back in history further than when you opened the model

Expected result: custom warning modal **is shown**.

**3. Custom warning modal is not shown when leaving editing model query with no changes**

1. Click "Edit query definition"
2. Press and hold back button in your browser to open up history menu
3. Go back in history further than when you opened the model

Expected result: custom warning modal **is not shown**.

**4. Custom warning modal is not shown when leaving editing model metadata with no changes**

1. Click "Edit metadata"
2. Press and hold back button in your browser to open up history menu
3. Go back in history further than when you opened the model

Expected result: custom warning modal **is not shown**.

**5. Custom warning modal is not shown when saving model query changes**

1. Click "Edit query definition"
2. Change something in the query definition
3. Click "Save changes"

Expected result: custom warning modal **is not shown**.

**6. Custom warning modal is not shown when saving model metadata changes**

1. Click "Edit metadata"
2. Change something in the metadata
3. Click "Save changes"

Expected result: custom warning modal **is not shown**.

**7. Custom warning modal is not shown when navigating between query/metadata tabs with unsaved changes and running the question**

1. Click "Edit metadata"
2. Change something in the metadata
3. Open "Query" tab
4. Change something in the query definition
5. Run the question

Expected result: custom warning modal **was not shown** at any stage.

### Demo

https://github.com/metabase/metabase/assets/6830683/2f55ab68-9a65-4751-8b8a-0fb9c517e257
